### PR TITLE
feat: add LangfuseProvider for remote trace evaluation

### PR DESCRIPTION
**Pull request by afarntrog**
### Summary

Adds `LangfuseProvider`, the first concrete implementation of the `TraceProvider` interface, enabling evaluation of Strands agent traces stored in [Langfuse](https://langfuse.com). This lets users run the evals pipeline against production agent sessions without needing local in-memory trace data.

### Motivation

The existing eval framework requires traces to be available in-process. Teams using Langfuse as their observability backend had no way to feed those traces into the evaluation pipeline. This PR closes that gap by fetching, paginating, and converting Langfuse observations into the typed span model (`InferenceSpan`, `ToolExecutionSpan`, `AgentInvocationSpan`) that evaluators already consume.

### What changed

**New: `LangfuseProvider`** (`src/strands_evals/providers/langfuse_provider.py` — 620 lines)
- Implements `TraceProvider.get_evaluation_data(session_id)` → `TaskOutput`
- Fetches traces by session ID with automatic pagination (`_fetch_all_pages`)
- Retries on `ReadTimeout` with exponential backoff via `tenacity` (`_call_with_retry`)
- Converts Langfuse observations to evals span types:
  - `generation` → `InferenceSpan` (with full message parsing: user/assistant, tool calls, tool results)
  - `span` with name `Tool: *` → `ToolExecutionSpan`
  - `span` with name `Agent *` → `AgentInvocationSpan`
- Extracts final agent output from the last trace's last inference span
- Configurable timeout, credentials via constructor args or env vars (`LANGFUSE_PUBLIC_KEY`, `LANGFUSE_SECRET_KEY`, `LANGFUSE_HOST`)

**Packaging**
- Added `langfuse` optional dependency group (`langfuse>=2.0.0`) in `pyproject.toml`
- Added `langfuse` to the hatch test environment features

**Cleanup**
- Removed redundant `pass` statements from exception classes
- Fixed missing newline at end of `__init__.py`

### How to try it

```python
from strands_evals.providers import LangfuseProvider

provider = LangfuseProvider(
    public_key="pk-lf-...",
    secret_key="sk-lf-...",
)
task_output = provider.get_evaluation_data(session_id="your-session-id")
# task_output.output  → final agent response string
# task_output.trajectory  → Session with typed traces/spans
```

---

## Related Issues

https://github.com/strands-agents/evals/issues/98

## Type of Change

New feature

## Testing

- **29 unit tests** covering all conversion paths, pagination, retry logic, error handling, and edge cases
- **11 integration tests** (`tests_integ/test_langfuse_provider.py`) that hit a real Langfuse instance

- [x] I ran `hatch run prepare`

## Checklist
- [x] I have read the CONTRIBUTING document
- [x] I have added any necessary tests that prove my fix is effective or my feature works
- [ ] I have updated the documentation accordingly
- [ ] I have added an appropriate example to the documentation to outline the feature, or no new docs are needed
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/strands_evals/providers/__init__.py [hunk 1]
```diff
@@ -14,4 +14,4 @@
     "TraceNotFoundError",
     "TraceProvider",
     "TraceProviderError",
-]
```
Comment (mkmeral) [inline src/strands_evals/providers/__init__.py:17]: don't we want the providers here? can we do lazy loading? https://github.com/strands-agents/sdk-python/blob/main/src/strands/experimental/bidi/__init__.py#L70

File: pyproject.toml [hunk 2]
```diff
@@ -44,6 +44,7 @@ dev = [
     "ruff>=0.13.0,<0.15.0",
 ]
 
+langfuse = ["langfuse>=2.0.0"]
```
Comment (mkmeral) [inline pyproject.toml]: can we add a ceiling for 3?

File: src/strands_evals/providers/langfuse_provider.py [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+    Session objects ready for the evaluation pipeline.
+    """
+
+    def __init__(
+        self,
+        public_key: str | None = None,
+        secret_key: str | None = None,
+        host: str | None = None,
+        timeout: int = _DEFAULT_TIMEOUT,
+    ):
```
Comment (mkmeral) [inline src/strands_evals/providers/langfuse_provider.py:59]: nit: docstring. i know it's kinda self explanatory, but it'll be also used to generate api ref in docs

File: src/strands_evals/providers/__init__.py (hunk 1 above)
Comment (afarntrog) [reply src/strands_evals/providers/__init__.py:17]: Good call. Updated!

File: pyproject.toml [hunk 4]
```diff
@@ -44,6 +44,7 @@ dev = [
     "ruff>=0.13.0,<0.15.0",
 ]
 
+langfuse = ["langfuse>=2.0.0,<3"]
```
Comment (mkmeral) [inline pyproject.toml:47]: smallest nit ever :p i'd do 3.0.0

File: src/strands_evals/providers/langfuse_provider.py [hunk 5]
```diff
@@ -0,0 +1,592 @@
...
+
+    Fetches traces and observations via the Langfuse Python SDK,
+    converts Langfuse observations to typed evals spans, and returns
+    Session objects ready for the evaluation pipeline.
+    """
+
+    def __init__(
+        self,
+        public_key: str | None = None,
+        secret_key: str | None = None,
```
Comment (poshinchen) [inline src/strands_evals/providers/langfuse_provider.py:56]: nit: we should support reading the public key and secret key via the env vars too (LANGFUSE_PUBLIC_KEY, LANGFUSE_SECRET_KEY, and LANGFUSE_BASE_URL) as a follow-up task

Comment (poshinchen) [reply src/strands_evals/providers/langfuse_provider.py:56]: ah nvm I see that at the bottom lol

File: src/strands_evals/providers/langfuse_provider.py [hunk 6]
```diff
@@ -0,0 +1,592 @@
...
+                f"Langfuse: traces found for session_id='{session_id}' but none contained convertible observations"
+            )
+
+        output = self._extract_output(session)
+
+        return TaskOutput(output=output, trajectory=session)
+
+    # --- Internal: fetching ---
+
+    def _fetch_all_pages(self, fetch_fn: Callable[..., Any], **kwargs: Any) -> list[Any]:
```
Comment (poshinchen) [inline src/strands_evals/providers/langfuse_provider.py:130]: hmm, can we scope down the return type? Or does Langfuse's API returns Any?

File: src/strands_evals/providers/langfuse_provider.py [hunk 7]
```diff
@@ -0,0 +1,592 @@
...
+            span_id=obs.id,
+            session_id=session_id,
+            parent_span_id=obs.parent_observation_id,
+            start_time=obs.start_time,
+            end_time=obs.end_time,
+        )
+
+    # --- Internal: conversion methods ---
+
+    def _convert_generation(self, obs: Any, session_id: str) -> InferenceSpan:
```
Comment (poshinchen) [inline src/strands_evals/providers/langfuse_provider.py:247]: Discussed offline, as a follow-up task we need to support either:
1. Integrate with our sessionMapper so we don't have to maintain the extraction logic in two places.
2. or support latest semantic convention's extraction which squashes the user.input / tool.input into input.messages.

File: src/strands_evals/providers/langfuse_provider.py (hunk 6 above)
Comment (afarntrog) [reply src/strands_evals/providers/langfuse_provider.py:130]: Discussed offline
